File: backend/app/api/routes/usage.py
"""用户端 Token 用量统计 API（仅统计当前登录用户）。"""
from __future__ import annotations

from fastapi import APIRouter, Depends

from app.core.user_auth import get_current_user
from app.db.database import get_db

router = APIRouter(prefix="/api/usage", tags=["usage"])


@router.get("")
async def recent_usage(limit: int = 200, user: dict = Depends(get_current_user)):
    db = get_db()
    rows = await db.execute_fetchall(
        "SELECT * FROM token_usage WHERE user_id=? ORDER BY id DESC LIMIT ?",
        (user["id"], min(limit, 1000)),
    )
    return [dict(r) for r in rows]


@router.get("/summary")
async def usage_summary(user: dict = Depends(get_current_user)):
    db = get_db()
    rows = await db.execute_fetchall(
        "SELECT call_type, COUNT(*) AS calls, SUM(prompt_tokens) AS prompt_tokens, "
        "SUM(completion_tokens) AS completion_tokens, SUM(total_tokens) AS total_tokens, "
        "AVG(latency_ms) AS avg_latency_ms "
        "FROM token_usage WHERE user_id=? GROUP BY call_type ORDER BY total_tokens DESC",
        (user["id"],),
    )
    total_row = await db.execute_fetchone(
        "SELECT COUNT(*) AS calls, COALESCE(SUM(total_tokens),0) AS total_tokens "
        "FROM token_usage WHERE user_id=?",
        (user["id"],),
    )
    return {
        "by_call_type": [dict(r) for r in rows],
        "total": {
            "calls": total_row["calls"] if total_row else 0,
            "total_tokens": total_row["total_tokens"] if total_row else 0,
        },
    }

File: frontend/src/api/usage.ts
import { api } from './http.ts';
import { userHeaders, adminHeaders } from './auth.ts';

export interface UsageRecord {
  id: number;
  call_type: string;
  model: string | null;
  user_id: string | null;
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
  latency_ms: number;
  created_at: string;
}

export interface UsageSummaryItem {
  call_type: string;
  calls: number;
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
  avg_latency_ms: number;
}

export interface UsageSummary {
  by_call_type: UsageSummaryItem[];
  total: { calls: number; total_tokens: number };
}

export interface AdminUsageRow {
  username: string;
  user_id: string | null;
  calls: number;
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
  last_active: string | null;
}

export interface AdminUsageItem {
  call_type: string;
  calls: number;
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
  avg_latency_ms: number | null;
}

export interface AdminUsageRecent {
  id: number;
  call_type: string;
  model: string | null;
  total_tokens: number;
  latency_ms: number | null;
  created_at: string;
  username: string;
}

export interface AdminUsage {
  by_user: AdminUsageRow[];
  by_call_type: AdminUsageItem[];
  total: { calls: number; total_tokens: number };
  recent: AdminUsageRecent[];
}

export const apiAdminUsage = {
  get: () => api<AdminUsage>('/api/admin/usage', { headers: adminHeaders() }),
};

// ---- Usage (用户端) ----
export const apiUsage = {
  summary: () => api<UsageSummary>('/api/usage/summary', { headers: userHeaders() }),
  recent: (limit = 200) =>
    api<UsageRecord[]>(`/api/usage?limit=${limit}`, { headers: userHeaders() }),
};
